### PR TITLE
Add report functionality to update step

### LINKED ISSUE #2968: Stop modifying local_ministep in make_update_data
Currently make_update_data is modifying the ministep by passing in the data for the observations used. This is bad for at least two reasons.
1. ministep is being modified quietly under the rug
2. ministep should only hold configuration for the local update and not hold actual data

We should avoid storing the data vectors and store only information regarding what observations was actually used. The modification of the ministep should also be moved into the forefront  of the code.

### PR DESCRIPTION
**Pull request by oyvindeide**
**Issue**
Resolves #2968 

This propagates information about the update step from the analysis module. More information is added at higher levels to get a better idea of what analysis has taken place, also related to: https://github.com/equinor/ert/issues/907

## Pre review checklist

- [x] Added appropriate labels

Adding labels helps the maintainers when writing release notes, see sections and the
corresponding labels here: https://github.com/equinor/ert/blob/main/.github/release.yml

### CODE REVIEW

File: libres/lib/include/ert/enkf/enkf_analysis.hpp [hunk 1]
```diff
@@ -27,9 +27,18 @@
 
 #include <ert/enkf/obs_data.hpp>
 
-void enkf_analysis_fprintf_obs_summary(const obs_data_type *obs_data,
-                                       const meas_data_type *meas_data,
-                                       const char *ministep_name, FILE *stream);
+class UpdateSnapshot {
```
Comment (ManInFez) [inline libres/lib/include/ert/enkf/enkf_analysis.hpp:30]: I would think that this class would belong to the analysis module?

File: res/enkf/es_update.py [hunk 2]
```diff
@@ -60,9 +60,10 @@ def analysis_smoother_update(
 
     _lib.update.copy_parameters(source_fs, target_fs, ensemble_config, ens_mask)
 
+    reports = []
```
Comment (ManInFez) [inline res/enkf/es_update.py:63]: There should be a link to which ministep it came from. The ministeps have names, so it should at least be a dict

File: res/enkf/es_update.py [hunk 3]
```diff
@@ -158,3 +160,5 @@ def smootherUpdate(self, run_context):
             source_fs,
             target_fs,
         )
+        self.ert.reports = update_reports
```
Comment (ManInFez) [inline res/enkf/es_update.py:163]: The reports should be appended and not overwritten, and  preferably linked to the current  update iteration

File: libres/lib/include/ert/enkf/enkf_analysis.hpp (hunk 1 above)
Comment (ManInFez) [reply libres/lib/include/ert/enkf/enkf_analysis.hpp:30]: Also, why is it called Snapshots here and referred to reports elsewhere?

File: libres/lib/include/ert/enkf/enkf_analysis.hpp [hunk 4]
```diff
@@ -27,9 +27,18 @@
...
+public:
+    std::vector<std::string> obs_name;
+    std::vector<double> obs_value;
+    std::vector<double> obs_std;
```
Comment (ManInFez) [inline libres/lib/include/ert/enkf/enkf_analysis.hpp:34]: should be error, not std

Comment (oyvindeide) [reply libres/lib/include/ert/enkf/enkf_analysis.hpp]: It is `std` everywhere else in the code, but it is of course not more correct with error

File: res/enkf/es_update.py (hunk 3 above)
Comment (oyvindeide) [reply res/enkf/es_update.py:163]: Yes, this is not the best solution, but not quite sure what the best is, so open to suggestions - did consider injecting more context higher in the callstack, for example configuration and target/source fs to make it unique. What do you think?

**Comment by jondequinor** [comment]
Says _WIP_ so I made it into a draft.

File: libres/lib/include/ert/analysis/update.hpp [hunk 5]
```diff
@@ -39,6 +42,7 @@ class update_data_type : public std::enable_shared_from_this<update_data_type> {
     Eigen::MatrixXd R;
     std::optional<Eigen::MatrixXd> A;
     std::vector<bool> obs_mask;
+    UpdateSnapshot update_report;
```
Comment (ManInFez) [inline libres/lib/include/ert/analysis/update.hpp:45]: is it a snapshot or report?

File: libres/lib/include/ert/enkf/enkf_analysis.hpp [hunk 6]
```diff
@@ -26,9 +26,29 @@
 
 #include <ert/enkf/obs_data.hpp>
 
-void enkf_analysis_fprintf_obs_summary(const obs_data_type *obs_data,
-                                       const meas_data_type *meas_data,
-                                       const char *ministep_name, FILE *stream);
+class UpdateSnapshot {
+    std::vector<std::string> obs_name_;
```
Comment (ManInFez) [inline libres/lib/include/ert/enkf/enkf_analysis.hpp:30]: make explicitly private

File: libres/lib/include/ert/enkf/enkf_analysis.hpp [hunk 7]
```diff
@@ -26,9 +26,29 @@
...
+    std::vector<double> obs_value_;
+    std::vector<double> obs_error_;
+    std::vector<std::string> obs_status_;
+    std::vector<double> response_mean_;
+    std::vector<double> response_std_;
+
+public:
+    const std::vector<std::string> &obs_name() const { return obs_name_; }
+    const std::vector<double> &obs_value() const { return obs_value_; }
```
Comment (ManInFez) [inline libres/lib/include/ert/enkf/enkf_analysis.hpp:39]: use get_ prefix for getter functions

File: res/enkf/es_update.py [hunk 8]
```diff
@@ -33,6 +37,17 @@ def config_is_correct(updatestep, iterable):
...
 
+@dataclass
+class SmootherSnapshot:
+    source_case: str
+    target_case: str
+    analyis_module: str
+    analysis_configuration: Dict[str, Any]
+    alpha: float
+    std_cutoff: float
+    ministeps: Dict[str, UpdateSnapshot] = field(default_factory=dict)
```
Comment (ManInFez) [inline res/enkf/es_update.py:48]: ministep_snapshots

File: ert_shared/libres_facade.py [hunk 9]
```diff
@@ -276,7 +276,7 @@ def gen_kw_priors(self):
         return self._enkf_main.getKeyManager().gen_kw_priors()
 
     def get_update_step(self):
```
Comment (ManInFez) [inline ert_shared/libres_facade.py:278]: This function no longer return an "update_step"

**Review by ManInFez — COMMENTED** [review]
This is a solid piece of work. Some smaller comments, and we should create an observation-milestone which iterates on top of this.

Some future work includes but are not limited to:
- extract all ministeps when exporting to ert-storage
- handle groups of observations in a more meaningful way

**Review by ManInFez — APPROVED** [review]

**Comment by ManInFez** [comment]
Remember to squash last commit

**Pull request merged**